cpp/arcticdb/storage/failure_simulation.hpp

```diff
@@ -7,56 +7,148 @@
+        proxy_(type);
+    }
+};
+
+inline std::ostream& operator<<(std::ostream& out, const FailureAction& action) {
+    return out << action.description_;
+}
+
+namespace action_factories { // To allow `using namespace`
+    static inline FailureAction no_op("no_op", [](FailureType){});
+
+    /** Raises the given exception with the given probability. */
+    template<class Exception = StorageException>
+    static inline FailureAction fault(double probability = 1.0) {
+        util::check_arg(probability >= 0, "Bad probability: {}", probability);
+
+        static FailureAction do_fault("raise", [](FailureType failure_type) {
+            throw Exception(fmt::format("Simulating {} storage failure", failure_type));
+        });
+
+        if (probability >= 1.0) {
+            return do_fault;
+        } else {
+            return {fmt::format("fault({})", probability),
+                    [prob=probability](FailureType failure_type) {
+                        thread_local std::once_flag flag;
+                        std::call_once(flag, [seed = uint64_t(&failure_type)]() { init_random(seed); });
+                        if (random_probability() < prob) {
+                            do_fault(failure_type);
+                        }
+                    }};
+        }
+    }
+}
+
+/** Independent state for each FailureType. Thread-safe except for the c'tors. */
+class FailureTypeState {
+public:
+    using ActionSequence = std::vector<FailureAction>;
+    static_assert(std::is_copy_assignable_v<ActionSequence>);
+
+private:
+    friend class StorageFailureSimulator;
+
+    const ActionSequence sequence_;
+    std::atomic<size_t> cursor_ {0}; // Index into sequence
+
+public:
+    FailureTypeState(ActionSequence sequence) :
+            sequence_(sequence.empty() ? ActionSequence{action_factories::no_op} : std::move(sequence)) {}
+
+    const ActionSequence::value_type& pick_action() {
+        if (cursor_ < sequence_.size()) {
+            if (auto local = cursor_.fetch_add(1); local < sequence_.size()) {
+                return sequence_[local];
+            }
+        }
+        return sequence_[sequence_.size() - 1];
+    }
 };
 
 class StorageFailureSimulator {
 public:
+    using ParamActionSequence = FailureTypeState::ActionSequence;
+    /**
+     * Easy-to-copy parameters that can be used to configure this class. Useful in parameterized tests.
+     * The string is a sequence of "action indicators" selecting the action to perform for each call.
+     * After this sequence is exhausted, the last action is used for all subsequent calls.
+     */
+    using Params = std::unordered_map<FailureType, ParamActionSequence>;
+
     static std::shared_ptr<StorageFailureSimulator> instance();
-    static std::shared_ptr<StorageFailureSimulator> instance_;
-    static std::once_flag init_flag_;
-    static void init(){
+    /* Clears/re-initializes the simulator. All action sequences, custom actions and states will be cleared. */
```

Comment:
The comment seems redundant.

Reply:
I like documenting behaviour that is not verbatim in the function body to indicate intent.

Reply:
I do agree that clear does pretty much exactly what you expect it to, shouldn't comments be for behaviour that is surprising? The issue is really consistency with the rest of the project, is it to be policy to do this everywhere from now on, or are we just doing it for this particular file?

Reply:
Will remove